src/lib.rs

```diff
@@ -191,145 +280,296 @@ where
+      first_secondary_circuit.clone(),
       pp.ro_consts_circuit_secondary.clone(),
     );
     let _ = circuit_secondary.synthesize(&mut cs_secondary);
     let (u_secondary, w_secondary) = cs_secondary
       .r1cs_instance_and_witness(&pp.r1cs_shape_secondary, &pp.r1cs_gens_secondary)
       .map_err(|_e| NovaError::UnSat)?;
 
+    let state = Self {
+      r_W_primary: RelaxedR1CSWitness::from_r1cs_witness(&pp.r1cs_shape_primary, &w_primary),
+      r_U_primary: RelaxedR1CSInstance::from_r1cs_instance(
+        &pp.r1cs_gens_primary,
+        &pp.r1cs_shape_primary,
+        &u_primary,
+      ),
+      l_w_primary: w_primary,
+      l_u_primary: u_primary,
+      r_W_secondary: RelaxedR1CSWitness::<G2>::default(&pp.r1cs_shape_secondary),
+      r_U_secondary: RelaxedR1CSInstance::<G2>::default(
+        &pp.r1cs_gens_secondary,
+        &pp.r1cs_shape_secondary,
+      ),
+      l_w_secondary: w_secondary,
+      l_u_secondary: u_secondary,
+      zn_primary: z1_primary,
+      zn_secondary: z1_secondary,
+      _p_c1: Default::default(),
+      _p_c2: Default::default(),
+      a1: Default::default(),
+      a2: Default::default(),
+      num_steps: 1,
+    };
+
+    Ok(state)
+  }
+
+  /// create a new `RecursiveSNARK` for circuits implementing StepCompute
+  pub fn prove<'a>(
+    pp: &PublicParams<G1, G2, C1, C2, A1, A2>,
+    num_steps: Option<usize>,
+    z0_primary: &IO<'a, G1::Scalar, A1>,
+    z0_secondary: &IO<'a, G2::Scalar, A2>,
+    c_primary: Option<C1>,
+    c_secondary: Option<C2>,
+  ) -> Result<Self, NovaError>
+  where
+    C1: StepCompute<'a, G1::Scalar, A1>,
+    C2: StepCompute<'a, G2::Scalar, A2>,
+  {
+    let c_primary = if let Some(c_primary) = c_primary {
+      c_primary
+    } else {
+      pp.c_primary.clone()
+    };
+    let c_secondary = if let Some(c_secondary) = c_secondary {
+      c_secondary
+    } else {
+      pp.c_secondary.clone()
+    };
+
+    if let Some(num_steps) = num_steps {
+      let mut primary_iterator = ComputeState {
+        circuit: c_primary,
+        val: z0_primary.clone(),
+      }
+      .take(num_steps);
+
+      let mut secondary_iterator = ComputeState {
+        circuit: c_secondary,
+        val: z0_secondary.clone(),
+      }
+      .take(num_steps);
+
+      Self::prove_with_iterators(
+        pp,
+        z0_primary,
+        z0_secondary,
+        &mut primary_iterator,
+        &mut secondary_iterator,
+      )
+    } else {
+      let mut primary_iterator = ComputeState {
+        circuit: c_primary,
+        val: z0_primary.clone(),
+      };
+
+      let mut secondary_iterator = ComputeState {
+        circuit: c_secondary,
+        val: z0_secondary.clone(),
+      };
+      Self::prove_with_iterators(
+        pp,
+        z0_primary,
+        z0_secondary,
+        &mut primary_iterator,
+        &mut secondary_iterator,
+      )
+    }
+  }
+
+  /// create a new `RecursiveSNARK` from primary and secondary iterators
+  pub fn prove_with_iterators(
```

Comment:
Would this need to be public?

Reply:
Changed.

Reply:
However, I've changed this one back. See answer to question below.